training/metrics.py

```diff
@@ -24,92 +21,244 @@ def count_trainable_params(model):
   """
   weights = model.trainable_weights
-  total_trainable_params = int(sum(np.prod(p.shape.as_list()) for p in object_identity.ObjectIdentitySet(weights)))
+  total_trainable_params = int(sum(np.prod(tf.size(p)) for p in object_identity.ObjectIdentitySet(weights)))
   return total_trainable_params
 
+def _count_linear_layer(layer, num_of_blades):
+  """
+  Note: This function calculates the flops for any linear layer
 
-def _count_flops_conv2d(layer):
-  if layer.data_format == "channels_first":
-    input_channels = layer.input_shape[1]
-    output_channels, h, w, = layer.output_shape[1:]
-  elif layer.data_format == "channels_last":
-    input_channels = layer.input_shape[3]
-    h, w, output_channels = layer.output_shape[1:]
-  w_h, w_w = layer.kernel_size
-
-  n_mult = h * w * output_channels * input_channels * w_h * w_w
-  n_add = w_h * w_w * input_channels * h * w * output_channels
+  Args:
+    layer (tf.keras.layer.Conv2D or DepthwiseConv2D or Dense): any linear layer from the keras model
+    num_of_blades: This value represent the number of blades we use for different upstride types.
 
-  flops = n_mult + n_add
+  Returns:
+      int : total FLOPs for the given layer.
+  """
+  input_shape = layer.input_shape
+  output_shape = layer.output_shape
+  output_group = np.prod(output_shape[1:])
+  class_name = layer.__class__.__name__
+
+  # we get output image size (H*W) (this is channel first/last dependent)
+  if len(input_shape) == 4: # 2D Conv and DepthWise
+    # get the channel index
+    channel_index = 1 if layer.data_format == "channels_first" else -1
+    # get the input channels and output channels
+    input_channels = input_shape[channel_index]
+    output_channels = output_shape[channel_index]
+
+  # we get kernel size (this is layer dependant (Dense vs Conv))
+  # Depthwise layer output channels are same as input.
+  # It is a generalized case of group convolution where input channels are same as number of groups
+  # The expected output channel is output_channels / groups. 
+
+  # When DepwiseConv2D is substituted with Conv2D with groups = input_channels. The division operation on the groups are
+  # already applied to the input channels. eg: H x W x (c_in / groups).
+  if class_name == "DepthwiseConv2D": 
+    # In keras, the layer.output_shape for DepthwiseConv2D is (h * w * c_out), where c_out is same as c_in.
+    # Hence dividing by input_channels to get the correct FLOPs count
+    output_group /= input_channels 
+  if len(input_shape) == 4:
+    kernel_size = np.prod(layer.kernel_size)
+    input_group = np.prod([kernel_size, input_channels])
+  elif len(input_shape) == 2: # Dense 
+    input_group = input_shape[1]  # (B, C)
+  else:
+    raise NotImplementedError("Flops for {layer.name} layer not implemented")
+
+  # compute the product groups
+  n_mul = (num_of_blades**2) * (input_group * output_group)
+  # max(1, num_of_blades-1) to ensure the n_add doesn't becomes zero when num_of_blades = 1
```

Comment:
I think it needs a better explanation than this. I understand the N*N-1 term : when you multiply N terms together, you have N-1 additions (although the best strategy is only log2(N)+1 for a sum accumulation), but for N=1 it is still unclear.

I guess there are two independent term that we add : the normal convolution (with N times the output filters, and N times the batch) and the accumulation of various blades multiplied together (which is (N-1)*N*output_size. We should differentiate these two terms, and then it is very clear that the second term is 0 when N=1

Reply:
For N=1, we haven't considered the -1 term as its negligible. I have added this to the doc string above.